File: 2025-09-September-GFG-POTD/September 12 - Minimize the Heights II/Solution.py
class Solution:
    def getMinDiff(self, arr, k):
        n = len(arr)
        arr.sort()

        ans = arr[-1] - arr[0]

        smallest = arr[0] + k
        largest = arr[-1] - k

        for i in range(n - 1):
            min_val = min(smallest, arr[i + 1] - k)
            max_val = max(largest, arr[i] + k)

            if min_val < 0:
                continue  # skip invalid cases

            ans = min(ans, max_val - min_val)

        return ans

File: 2025-09-September-GFG-POTD/September 12 - Minimize the Heights II/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int getMinDiff(int arr[], int n, int k) {
        sort(arr, arr + n);

        int ans = arr[n - 1] - arr[0];

        int smallest = arr[0] + k;
        int largest = arr[n - 1] - k;

        for (int i = 0; i < n - 1; i++) {
            int minVal = min(smallest, arr[i + 1] - k);
            int maxVal = max(largest, arr[i] + k);

            if (minVal < 0) continue; // skip invalid cases

            ans = min(ans, maxVal - minVal);
        }

        return ans;
    }
};
